### DISCUSSION TITLE
Display average speed

### DISCUSSION THREAD

**Discussion by DolomitesTom** [devemux86/cruiser#7] [Q&A] [answered]
Is it somehow possible to show the average speed (actually calculated average speed since start of the tour without pauses... stops should not decrease the average speed... pure driving average speed) in an additional optional info box on the right side? 

Is it furthermore possible to show this as a minimal statistics after the tour, somehow stored or calculated in the recorded track data? Could here possibly be distinguished between 2 values, once the average speed with and once without breaks?

**Comment by lucvl5** [comment]
Hello, 
Such an add on is probably requested by more people but personally I would prefer the developers to focus on navigation specific requests, not on requests which can be fulfilled by other apps. I'm sure for getting this statistic a bunch of apps are available, with probably much more info than just the average speed, which is provided by most motorbikes anyways. As an example of what I see as a navigation request I just entered as a new discussion. This improvement avoids that you have to stop your trip to check for gas stations. 
Just my thought of course.

**Reply by 0709wiwiwi** [reply]
The motards are obviously more present here, but the simple Cruiser app has already been of interest to me as a cyclist for some time. I am following new proposals and looking at the impact for cycling navigation. The avg speed shown will not be too bad if set up properly, but sports cyclists are going to want to expand Tom's proposal possibly. I myself am not too keen on such statistics through only gps data. A summary of my short bike ride on 3 January 2023, and no, I don't check nor parse such (max speed) nonsense anymore.
![image](https://user-images.githubusercontent.com/54167201/210621294-7a7fd70e-f721-4042-8feb-f6809a48fa8c.png)
Notice that I got through the year change very well and still have quite a bit of explosiveness in my legs despite my 69 years. 
And all this powered by just 1 hp (human power).

**Reply by devemux86** [reply]
Something detailed like that would be nice.

Also a chart to visualize statistics / routing.

**Reply by 0709wiwiwi** [reply]
Provided misses like the max speed notice in the example can be avoided, the sports enthusiasts will quite appreciate this.   
![image](https://user-images.githubusercontent.com/54167201/210628003-4158c40f-e72f-487e-8232-099b1a600a29.png)
Speed data supplied by a real speed sensor obviously does not record such.  For walkers and runners is of course less obvious.

**Reply by devemux86** [reply]
Is it much simpler to add the maximum speed if you need it?

**Comment by devemux86** [comment]
> Is it somehow possible to show the average speed

Thanks for the suggestion, it is already in my list for implementation. :slightly_smiling_face: 

> in an additional optional info box on the right side

Navigation panels were created for this purpose: _add optional panels with additional information_.

> somehow stored or calculated in the recorded track data?

You mean the GPS recorded track? So average speed in navigation and from a GPS recorded track.

This can also work, a recorded track stores coordinates with additional info, like current speed, time, elevation, ...
Just like we can calculate the average speed during navigation, we can also calculate it from GPS recorded data.

> Could here possibly be distinguished between 2 values, once the average speed with and once without breaks?

Interesting idea, I need to think about the implementation.

> the developers

1 developer! :slightly_smiling_face: 

I am available for all suggestions, we just need to organize them properly, so that they can eventually be implemented.

**Comment by DolomitesTom** [comment]
> You mean the GPS recorded track? So average speed in navigation and from a GPS recorded track.

Yes, that's how it was meant.

> 1 developer! 🙂...
> I am available for all suggestions, we just need to organize them properly, so that they can eventually be implemented.

I know... perfect !

**Comment by WolfgangZ42** [comment]
> average speed, which is provided by most motorbikes anyways.

As lucvl5 wrote - while navigating you have your motorcycle board computer for this. And after the tour you may use the recorded track and gps-analysing tools for any statistics.

**Reply by DolomitesTom** [reply]
This is absolutely not comparable. Such a display doesn't differ between average speed with and without pauses. I don't use the apps from the motorbike manufacturers, neither from BMW nor from KTM and I have also older bikes with no dashboard at all. I only use Cruiser on every motorbike. That's the reason why I would like to have this information at one place, in Cruiser. I don't want to use other apps for statistics. I don't need other statistics than the 2 mentioned variants of average speed as minimal statistics.

**Reply by lucvl5** [reply]
You're right, but I use the app to plan and navigate. So please forgive me to request higher priority to navigation and planning related developments instead to options where alternatives are available.

**Reply by 0709wiwiwi** [reply]
And an additional game changer. Auto pause reliability in track records is not flawless. If you don't manually pause, have a coffee somewhere, your moving avg speed is quickly misrepresented.

**Reply by devemux86** [reply]
> Auto pause reliability in track records is not flawless.

GPS recording has a min speed filter (1-10 km/h), default is off.
Doesn't it help? Or it needs to be default on and what speed?

**Reply by 0709wiwiwi** [reply]
The fast bike boys and girls have the advantage that the record condition can be set less critically. The much slower walker cannot. Not a big deal as long as you realise that depending where exactly you paused the avg result can deviate. 'Experience by other app'. As a cyclist, I do trust the small bike computer through wheel sensor.

**Reply by lucvl5** [reply]
Maybe a stupid question, but why are people interested in the average speed? Especially as compared to the distance to the next gas station?
Asking for a friend 😀

**Comment by DolomitesTom** [comment]
I guess we could play with the minimal gps speed on recording. No need to change the default. 
But another problem: Sometimes when pausing the navigation, the resume button doesn't work anymore. I had the same effekt also in Kurviger 2. It happens quite often. Only stopping the navigation and starting new from the actual position helped then.Would this have an effect on the average speed of the tour (daytrip)?

**Comment by devemux86** [comment]
> Sometimes when pausing the navigation, the resume button doesn't work anymore.

Can you write more details, what happens on screen?
Is there a video available?

If auto hide buttons is on, you could not press it because it hides or something else?

> Would this have an effect on the average speed of the tour (daytrip)?

Start / restart navigation usually means a new process.
For statistical data we have to think about it and see how it will work with navigation states.

**Reply by DolomitesTom** [reply]
```
> Can you write more details, what happens on screen?
> Is there a video available?

Nothing happens on screen. It is just not possible to resume the navigation, nothing happens pressing the resume/navigation buttons, no overlaying by other app. I wonder why I have this problem alone, because I have it pretty often. I try to make a video when it happens.

A video doesn't exist because I would need then 2 smartphones. 
```

**Reply by devemux86** [reply]
> Nothing happens on screen.

Does the navigation button change back to the active icon, i.e. blue without the `\` line?
Does your GPS works? Is there a location symbol on the map and near the route?
Do the navigation panels at the top appear filled or empty without data?
Or is there something else?
I need more details to understand what the "not possible to resume" means. :slightly_smiling_face:

**Reply by 0709wiwiwi** [reply]
@DolomitesTom 
Hi Tom. A13 ? Pause: Do you take the  phone away from the bike ? Is the display constantly on or not ?

**Comment by WolfgangZ42** [comment]
https://forum.kurviger.de/t/app-display-average-speed/5551
You're asking (and waiting) nearly two years for it - in between you could have a solution with another App?
Btw: no-one supported this idea, as far as I know.

**Reply by DolomitesTom** [reply]
Can you please take your hateful comments somewhere else? Or just be quiet if you don't care about the topic?

Read the privacy policy of the other app and you'll know why I don't want to use it.

**Reply by devemux86** [reply]
Please everyone stay calm! :slightly_smiling_face: 

You will not find another app developer to discuss with users this much!
Everything is possible, I just need to organize well, features and priorities.

**Comment by DolomitesTom** [comment]
Today I had again this hanging in the app. It looks like this

![Screenshot_20230107-152712_Cruiser](https://user-images.githubusercontent.com/121371629/211160021-746b23c2-6f3a-4901-b7fe-cf1ef3dbe4fc.jpg)
You can press which ever button you want, no one is working. The more you push on the button the colors of the left 3 Buttons change in: sometimes all 3 are blue sometimes only one in the middle sometimes the two at the edge are blue. The only thing that you can do is using the back key to close the app. When you start it again, all works normal.

**Reply by devemux86** [reply]
Thanks for the extra details, indeed this is strange.
Pressed buttons must be blue and only 1 button can be pressed / blue at a time.

Are you sure the phone is working properly at that time without overheating or something?
Are there other applications - especially with overlays - active at the same time?

**Reply by DolomitesTom** [reply]
Yes the phone is working fine for sure. I use one overlaying app for traffic warnings.... the one with the green arrow.

**Comment by devemux86** [comment]
_A kind request: this topic is about the feature of average speed.
Please create new topics to discuss different things separately._

**Comment by DolomitesTom** [comment]
Now that Cruiser has such a phenomenal new menu design, the appetite comes naturally when eating. Therefore I would like to ask this feature "display average speed with and without pauses" (as well as storing these values maybe in the history of last tours or in the recorded track?) is on which position of the list :-) ?

**Reply by devemux86** [reply]
Indeed, the new menu design gives us the flexibility to offer easily more features.

Don't worry, I already have the average speed in my implementation list. :slightly_smiling_face: 

We need to see how route / track details can be stored and presented (in a graph?), combined with:
- https://github.com/devemux86/cruiser/discussions/35
- https://github.com/devemux86/cruiser/discussions/81

**Comment by DolomitesTom** [comment]
> Don't worry, I already have the average speed in my implementation list. 🙂

Just a question, no pressure: we will soon start with multi-day tours. This would be a good opportunity to test the average speed feature. How far away is a first implementation?

**Reply by devemux86** [reply]
@DolomitesTom thanks for the reminder!

A first implementation could be to display it as extra panel while navigation / follow are active.

Can you explain more the difference with the breaks?
In navigation a break is considered when we press "pause".

---

I am working on several new features for Cruiser and some are quite large, so I try to balance it all! :slightly_smiling_face:

**Comment by DolomitesTom** [comment]
> A first implementation could be to display it as extra panel while navigation / follow are active.

This was already great.

> Can you explain more the difference with the breaks?
In navigation a break is considered when we press "pause".

With breaks I mean pauses/stopps for lunch or drinking. Pausing the Cruiser app is one thing which interrupts the pure driving time (2nd average value). 1st timer could count further to result in the average including pauses. What happens if I do not pause the navigation, but completely stop and the resume later by starting the app again? Would then the last average speed be initialized again or restored? I would prefer restore.

The average speed could be the kilometers driven so far divided by the total time required so far including breaks (1st average value) or the kilometers driven so far related to the pure driving time, i.e. without stops and breaks (2nd average value). Both values are interesting to me.

**Reply by devemux86** [reply]
> What happens if I do not pause the navigation, but completely stop and the resume later by starting the app again?

Start navigation is now considered a "new" navigation, although can start anywhere on the route.
Keeping navigation variables between app restarts is a new feature that will need implementation.

Also, do you mean we allow both values to appear or select one of them with a setting
or switch with tap (like the remaining time / arrival time switch)?

**Reply by DolomitesTom** [reply]
Select one of them configurable would be sufficient to my point of view. Tapping on the figure could switch between the both. I think that's nice.

**Comment by devemux86** [comment]
Next Cruiser will display the average speed in navigation / follow.

Some things for discussion:
- **Calculation**:
  - Mean value of GPS speeds?
  - Or distance / time?
- **Life cycle**:
  - A simple implementation is to start the calculation when we start the navigation / follow.
  - Calculate the average until we stop navigation / follow or while the app is running (start / stop many times)?
- **Threshold**:
  - Calculate the average speed (by any method) if speed > threshold?
So average speed does not change much at traffic lights, breaks...
  - "Settings | Location | Advanced | Orientation threshold" speed option already exists.
- **Icon**: can you think of a better one? (look [here](https://fonts.google.com/icons?selected=Material+Icons))

<img src="https://user-images.githubusercontent.com/3484020/232826426-e459f728-dbf2-4c0c-98d6-9f8325e5f428.jpg" width="50%"/>

**Reply by DolomitesTom** [reply]
😁... am pleased ...
start with the simple implementation i fine

**Icon**
There is a special avg. pace symbol in google fonts.
![Screenshot_20230419-064834_Firefox](https://user-images.githubusercontent.com/121371629/232969847-00a5ed6f-92d2-4ade-ae91-81ea63f367aa.jpg)

https://fonts.google.com/icons?icon.query=average

**calculation**
I would say mean GPS should do.

speed > threshold is also fine. Threshold can be set to 0 to include pauses or to a different value to exclude pauses. Later it would be interesting to have both values (distance/time includes all pauses) in track details, but this is already in your list I guess.
 
I've asked myself already what other meaning than gps recording minimal speed this has:
Settings | Location | Advanced | Orientation threshold" speed option

**Reply by devemux86** [reply]
> I've asked myself already what other meaning than gps recording minimal speed this has:
Settings | Location | Advanced | Orientation threshold" speed option

There are several use cases:

- Use the GPS direction only when the GPS speed > selected threshold.
So the map does not rotate constantly when we stop, like at traffic lights.

- There is also the "Settings | Navigation | Recalculate route | In case of reverse direction".
Where the check for wrong direction is again performed when speed > selected threshold.

An adjustable filter is needed, because at very low or zero speeds the GPS sensor can report the wrong direction.

**Reply by DolomitesTom** [reply]
understand thanks

**Comment by ramondevic** [comment]
Can the calculation of the average speed be deactivated so as not to disturb the user and avoid cpu and battery consumption if this calculation is of no interest?

**Reply by devemux86** [reply]
The panels that are not enabled in navigation / follow do not activate their calculation processes.

_Cruiser has always been highly optimized for stability, speed, precise location and battery life._ :slightly_smiling_face:

**Reply by ramondevic** [reply]
Great !
This is the most important.

**Comment by ramondevic** [comment]
Perhaps it would be interesting that when you save the track made, there is the option to see the details of the track so that you can get as much information as possible... for example:

![track detalls 1](https://user-images.githubusercontent.com/122691401/232991648-14ccec38-825a-4faa-8cc2-16362bf300c3.jpg)
![track detalls 2](https://user-images.githubusercontent.com/122691401/232991687-fdb45356-45fa-4eb5-85fb-522854d1c8cb.jpg)

**Reply by WolfgangZ42** [reply]
we discussed this already in January ...

**Reply by devemux86** [reply]
@ramondevic this is planned for the future! :slightly_smiling_face:

**Reply by ramondevic** [reply]
The screenshots that I have put are taken from a program that I use to prepare the routes on foot or with a motorcycle (off) and are only as an example that many programs or applications already do this work and that might not be necessary in Cruiser .

I think that the analysis of the routes done is better done at home once the route is done and that it is not necessary to have it online while we are riditing.

Well, that's my opinion.

**Reply by DolomitesTom** [reply]
I prefer to have the relevant information all in cruiser. I don't want to use another App. I don't do analysis after the tour.

**Reply by devemux86** [reply]
We can start by offering the average speed panel in navigation / follow.

The statistics of the traveled route or the recorded track can be applied later. :slightly_smiling_face:

**Reply by lucvl5** [reply]
I also would like everything in cruiser because i want to change my planning during the trip. I don't carry a laptop on my bike...

**Comment by ramondevic** [comment]
This is what TwoNav Land shows

**Comment by devemux86** [comment, marked as answer]
Display of average speed is available in [Cruiser 3.0.18](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

Available in "Settings | Navigation | Display | Average speed" or search settings for "average".

It measures the average speed while navigation / follow is active while the app is running.

Please let me know how you see it!

<img src="https://user-images.githubusercontent.com/3484020/233592727-02fa70a6-89e2-457c-903f-6594d0a35dd6.jpg" width="50%"/><img src="https://user-images.githubusercontent.com/3484020/232826426-e459f728-dbf2-4c0c-98d6-9f8325e5f428.jpg" width="50%"/>

**Comment by DolomitesTom** [comment]
Hotly desired... now we have to wait again long time for Google 🙄

**Reply by DolomitesTom** [reply]
Works well while driving 👍. Which method is it now? Track details have disappeared, right? Max.Speed would also be good in trackdetails when ready.

**Reply by devemux86** [reply]
> Which method is it now?

- Calculates mean value of GPS speeds
- Uses speeds > "Settings | Location | Advanced | Orientation threshold" speed
So average speed does not change much at traffic lights, breaks...
- Works when navigation is enabled (not paused)
- Or when follow mode is enabled
- It retains the value until we exit the application

Track details are indeed planned for the future. :slightly_smiling_face:

**Comment by DolomitesTom** [comment]
Perfect... already valuable information driven out for me:
50 km known rural area, some village crossings, many curves 68 km/h average on rain-wet road, same distance return roads partly dried 78 km/h, on dry roads probably 86 km/h. 
If you know the area, you can adjust the average speed before each tour and always have precisely planned arrival times... that's what I like.

When the average could be kept when exiting and restarting the app until the track is deleted that would be fine. In track details for aftertour checks see addtional the average speed including all pauses, max speed, departure time, arrival time, break times (and some other information when easily accessible)?

**Reply by devemux86** [reply]
@DolomitesTom thanks for the quick testing!

Route / track details of various statistics are already planned.

As we can navigate a route without recording anything, which one do you mean?
- Navigation statistics
These are currently not saved: average speed resets when app (re)starts
- Recorded track statistics

**Comment by DolomitesTom** [comment]
Navigation statistics

That's what I don't know how you plan to realize the statistics. I thought you would save the data together with the track, but you are right, you can deactivate recording. Does it mean you save it differently? Then the question is: how do you reset the data at some point or maybe want to save it somehow with a toure name before resetting. Don't know what ideas do you have in mind.

**Reply by devemux86** [reply]
Navigation and GPS recording can work together or separately:
- We can navigate a route without recording a track
- We can record a track with follow mode without route

So it is indeed still under discussion how such statistics will be maintained and presented.

**Comment by DolomitesTom** [comment]
> So it is indeed still under discussion how such statistics will be maintained and presented.

just implement what you have in mind, should suffice as a basis for discussion, more ideas will come when using it :-)

**Comment by DolomitesTom** [comment]
after 1200km through Voges and Black Forrest I need to add something here. The saving of the track details data including the average speed with and without pauses need to get saved when exiting the app and restored when resuming the trip. This because the battery consumes power and gets warm, even when just pausing the tour and putting cruiser in the background. Therefore then normal handling would be to exit cruiser when pausing and starting again after the pause.

**Reply by devemux86** [reply]
@DolomitesTom indeed, a more permanent solution would be useful.

Do you mean the average speed calculated during navigation / follow?
Or something more?

The recorded track with its GPS data is always saved.
(I have not completed the track detail screen yet)

**Reply by DolomitesTom** [reply]
> Do you mean the average speed calculated during navigation / follow?

Yes mainly this. If you have some other useful statistics in mind, sure, why not.

**Comment by airald** [comment]
This option is very useful for me, but should be interesting to reset it by tapping over the value and maybe a OK/Cancel popup.

**Reply by devemux86** [reply]
> This option is very useful for me

:+1: 

> should be interesting to reset it by tapping over the value and maybe a OK/Cancel popup.

Average speed currently is not permanent, when Cruiser restarts it resets.

Or do you mean to reset it inside a running navigation?

**Reply by airald** [reply]
Yes, I mean to reset while navigating.

**Reply by devemux86** [reply]
@airald 

> reset while navigating

This should be easy, I could add such a dialog.

**Reply by devemux86** [reply]
> reset it by tapping over the value and maybe a OK/Cancel popup

It will be available this way in the next Cruiser update.

**Reply by devemux86** [reply]
Reset of average speed (tap its panel) is available in [Cruiser 3.0.23](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Reply by DolomitesTom** [reply]
I don't know why this reset was requested. Does it resets the average to initial? Or does it reset to the previous average when leaving and reopening the app?

**Reply by devemux86** [reply]
The average of any value starts with zero, so the reset initializes the average back to zero.

It may be useful if you do not want to restart the app to measure the average speed again.

**Comment by lucvl5** [comment]
I'm lost. Why are we even discussing average speed? Who cares? Seeing the route you actually did, ok. But average speed? Why not focus on the planning? Curvy versus fast is much more interesting for a planning tool, in my opinion.

**Reply by devemux86** [reply]
Everyone, no problem!

Smaller features can be easier to implement, some only take a few minutes!
You can be sure that larger features are being worked in the background. :slightly_smiling_face:

**Comment by DolomitesTom** [comment]
Average speed is very important for planning. It enables an exact estimation when arrive at which WP.
